Implement Kubernetes HA mode

bernardjkim opened the pull request:
## Description
This PR enables Kubernetes HA mode. 

A `--high-availability` boolean flag has been added to the planet `start` and `agent` commands. Default value will be read from container-environment variable `KUBE_HIGH_AVAILABILITY`.

If planet is running in HA mode, Kubernetes control plane components(`apiserver`, `controller-manager`, `scheduler`) will run on all master nodes vs running only on the elected leader.

bernardjkim committed:
Add --high-availability flags

bernardjkim committed:
Implement Kubernetes HA mode

If planet is running in HA mode, Kubernetes control plane components
will run on all master nodes.

bernardjkim committed:
Explicitly set --endpoint-reconciler-type and --apiserver-count

knisbet commented on tool/planet/agent.go:
```diff
@@ -143,85 +121,73 @@ func startLeaderClient(config agentConfig, agent agent.Agent, errorC chan error)
+	return client, nil
+}
+
+// manageParticipation starts voter participation when the election mode is
```
nitpick: referring to this as election mode here and above reads a bit weird to me. I would consider rewording this a bit to just say whether elections are enabled/disabled.

knisbet reviewed (commented):
I suspect we should shutdown the control plane when the elections key for a particular node is disabled. See the note inline below about the particular issue this would help with.

bernardjkim committed:
Start/Stop units when election is enabled/disabled

bernardjkim commented on tool/planet/agent.go:
```diff
@@ -231,29 +197,72 @@ func startLeaderClient(config agentConfig, agent agent.Agent, errorC chan error)
 
+			// Control plane components run on all master nodes if HA mode is enabled.
+			if conf.HighAvailability {
+				return
```
Updated in 495b0f2

While running with HA mode enabled , control plane units will be stopped when election is disabled, and units will start back up when election is enabled.

Behavior should be unchanged if running with HA mode disabled.

a-palchikov commented on tool/planet/start.go:
```diff
@@ -533,6 +534,12 @@ func addComponentOptions(config *Config) error {
 			fmt.Sprintf("--service-node-port-range=%v", config.ServiceNodePortRange))
 	}
+	if config.HighAvailability {
+		config.Env.Append(EnvAPIServerOptions, fmt.Sprintf("--endpoint-reconciler-type=lease"))
```
```suggestion
		config.Env.Append(EnvAPIServerOptions, "--endpoint-reconciler-type=lease")
```
ditto below

a-palchikov commented on tool/planet/agent.go:
```diff
@@ -143,85 +121,75 @@ func startLeaderClient(config agentConfig, agent agent.Agent, errorC chan error)
-				log.Errorf("failed to add voter for %v: %v", conf.PublicIP, trace.DebugReport(err))
+			err := client.AddVoter(ctx, config.leader.LeaderKey, config.leader.PublicIP, config.leader.Term)
+			if err != nil {
+				log.Errorf("failed to add voter for %v: %v", config.leader.PublicIP, trace.DebugReport(err))
```
```suggestion
				log.WithError(err),Errorf("Failed to add voter for %v.", config.leader.PublicIP)
```

a-palchikov replied:
`cancelVoter()` as this is exiting after that

a-palchikov commented on tool/planet/agent.go:
```diff
@@ -143,85 +121,75 @@ func startLeaderClient(config agentConfig, agent agent.Agent, errorC chan error)
+
+			// While running with HA enabled, start units when election is re-enabled
+			if config.leader.HighAvailability {
+				if err := startUnits(context.TODO()); err != nil {
```
Use `ctx` from above

a-palchikov reviewed (approved):
I would also describe the mode of operation even briefly - e.g. that it will run local agent elections  and the coredns.hosts will contain just the elected node as apiserver.

a-palchikov commented:
One thing I did not notice is how the failed api server handled on the leader node (according to agent leader status)? Will it cause failover of the agent leader and updates to coredns.hosts file?

knisbet reviewed (approved):
lgtm with Dima's suggestion for removing the unneeded fmt.Sprintf's in a couple of locations.

bernardjkim commented:
@a-palchikov
If I understood your question correctly, there doesn't seem to be any failover in place if the apiserver fails. Would it make sense to watch the apiserver on the leader and run re-election in case the apiserver fails?

a-palchikov commented:
@bernardjkim Yes, I see this as a possible weak link. In case the apiserver fails on the leading agent node in HA mode, there will be no DNS update as long as the leader stays the same.

bernardjkim committed:
Address feedback

- No need to Sprintf
- Log WithError
- cancelVoter if failed to add voter
- Use voter ctx when starting units

bernardjkim commented:
I'll go ahead and merge this PR for now. I'll work on implementing apiserver failover in a separate PR.
